nmigen/hdl/ast.py

```diff
@@ -1280,6 +1284,51 @@ def _rhs_signals(self):
         return self._lazy_lower()._rhs_signals()
 
 
+class ValueCastable:
+    """Base class for classes which can be cast to Values.
+
+    A ``ValueCastable`` can be cast to ``Value``, meaning its recise representation does not have
+    to be immediately known. This is useful in certain metaprogramming scenarios. Instead of
+    providing fixed semantics upfront, it is kept abstract for as long as possible, only being
+    cast to a concrete nMigen value when required.
+
+    Note that it is necessary to ensure that nMigen's view of representation of all values stays 
+    internally consistent. The class deriving from ``ValueCastable`` must decorate the ``as_value``
+    method with the ``lowermethod`` decorator, which ensures that all calls to ``as_value``return the
+    same ``Value`` representation. If the class deriving from ``ValueCastable`` is mutable, it is
+    up to the user to ensure that it is not mutated in a way that changes its representation after
+    the first call to ``as_value``.
+    """
+    def __new__(cls, *args, src_loc_at=0, **kwargs):
+        self = super().__new__(cls)
```

Comment:
Should this `__new__` method forward `*args, **kwargs`? (I'm not entirely sure, I would have to look up this part of the Python object model, but this is what I'd expect.)

Reply:
This section is slightly magic to me (overriding `__new__`, and especially the assignment of `self`). I agree with you that that makes more sense.

Reply:
`self` is an ordinary variable in Python with no special meaning in any context whatsoever. (What *does* have special meaning is the very first argument of a class method, but it has nothing to do with its *name*.)

Reply:
As for `__new__` I will double-check whether this is correct, so don't worry about that.

Reply:
Except it's an error (`TypeError: object() takes no parameters`). So I guess not.

Reply:
For the onlookers: [this](https://hg.python.org/cpython/file/78d36d54391c/Objects/typeobject.c#l3034) is the reason for the strange behavior of `__new__` here.